packages/grid/x-data-grid-pro/src/hooks/features/lazyLoader/useGridLazyLoader.ts

```diff
@@ -0,0 +1,140 @@
+import * as React from 'react';
+import {
+  useGridApiEventHandler,
+  GridFeatureModeConstant,
+  GridRenderedRowsIntervalChangeParams,
+  useGridSelector,
+  GridRowId,
+  gridVisibleSortedRowIdsSelector,
+  gridSortModelSelector,
+  gridFilterModelSelector,
+  useGridApiOptionHandler,
+  GridEventListener,
+} from '@mui/x-data-grid';
+import { useGridVisibleRows } from '@mui/x-data-grid/internals';
+import { getRenderableIndexes } from '@mui/x-data-grid/hooks/features/virtualization/useGridVirtualScroller';
+import { GridApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import { GRID_SKELETON_ROW_ROOT_ID } from './useGridLazyLoaderPreProcessors';
+import { GridFetchRowsParams } from '../../../models/gridFetchRowsParams';
+
+/**
+ * @requires useGridRows (state)
+ * @requires useGridPagination (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridLazyLoader = (
+  apiRef: React.MutableRefObject<GridApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    'onFetchRows' | 'rowsLoadingMode' | 'pagination' | 'paginationMode' | 'rowBuffer'
+  >,
+): void => {
+  const visibleRows = useGridVisibleRows(apiRef, props);
+  const rowIds = useGridSelector(apiRef, gridVisibleSortedRowIdsSelector);
+  const sortModel = useGridSelector(apiRef, gridSortModelSelector);
+  const filterModel = useGridSelector(apiRef, gridFilterModelSelector);
+  const renderedRowsIntervalCache = React.useRef<GridRenderedRowsIntervalChangeParams>({
+    firstRowToRender: 0,
+    lastRowToRender: 0,
+  });
+
+  const handleRenderedRowsIntervalChange = React.useCallback(
+    (params: GridRenderedRowsIntervalChangeParams) => {
+      const dimensions = apiRef.current.getRootDimensions();
+
+      if (
+        !dimensions ||
+        props.rowsLoadingMode !== GridFeatureModeConstant.server ||
+        renderedRowsIntervalCache.current.firstRowToRender === params.firstRowToRender ||
+        renderedRowsIntervalCache.current.lastRowToRender === params.lastRowToRender
+      ) {
+        return;
+      }
+
+      const renderedRowsIds: Array<GridRowId | null> = [...rowIds].splice(
+        params.firstRowToRender,
+        params.lastRowToRender - params.firstRowToRender,
+      );
+      const hasSkeletonRowIds = renderedRowsIds.some(
+        (rowId) => `${rowId}`.indexOf(GRID_SKELETON_ROW_ROOT_ID) >= 0,
+      );
+
+      if (!hasSkeletonRowIds) {
+        return;
+      }
+
+      renderedRowsIntervalCache.current = params;
+      const fetchRowsParams: GridFetchRowsParams = {
+        firstRowToRender: params.firstRowToRender,
```

Comment:
I don't know if you plan to improve this, but the `firstRowToRender` might be the index of a row already loaded, making the server to return data not needed. I think we should pass to the callback only the indexes, between `firstRowToRender` and `lastRowToRender`, that contains unloaded rows.

Adding `rowBuffer=0` and `rowThresold=0` to a demo makes easier to see this problem.